### PR TITLE
Modularized Bert2LSTM

### PR DESCRIPTION
**Pull request by avishreekh**
Solves #57

### CODE REVIEW

File: KD_Lib/KD/text/Bert2Lstm/bert2lstm.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+            distill_val_loader,
+            optimizer_teacher,
+            None,
+            loss_fn,
+            temp,
+            distil_weight,
+            device,
+            log,
+            logdir,
+        )
```
Comment (Het-Shah) [inline KD_Lib/KD/text/Bert2Lstm/bert2lstm.py:84]: Please shift this to the beginning of the __init__ method this is the proper flow of calling the base class

File: KD_Lib/KD/text/Bert2Lstm/bert2lstm.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    def calculate_kd_loss(self, y_pred_student, y_pred_teacher, y_true):
+        """
+        Function used for calculating the KD loss during distillation
+        :param y_pred_student (torch.FloatTensor): Prediction made by the student model
+        :param y_pred_teacher (torch.FloatTensor): Prediction made by the teacher model 
+        :param y_true (torch.FloatTensor): Original label
+        """
+
+        # soft_teacher_out = F.softmax(y_pred_teacher / self.temp, dim=0)
+        # soft_student_out = F.log_softmax(y_pred_student / self.temp, dim=0)
```
Comment (Het-Shah) [inline KD_Lib/KD/text/Bert2Lstm/bert2lstm.py:115]: why are these two commented out? In the original loss function we take T-softened probablity

File: KD_Lib/KD/text/Bert2Lstm/bert2lstm.py (hunk 1 above)
Comment (avishreekh) [reply KD_Lib/KD/text/Bert2Lstm/bert2lstm.py:84]: Thank you for the detailed review. 
1. I am not very familiar with tensorboard and hence did not include it here. I would request someone else to possibly take this up or maybe I can do it later.
2. That seems like a nice approach. Are you suggesting that we have a class for the model which handles its training/distillation, etc.? 
3. I will change that. Thank you for pointing that out.

File: KD_Lib/KD/text/Bert2Lstm/bert2lstm.py (hunk 2 above)
Comment (avishreekh) [reply KD_Lib/KD/text/Bert2Lstm/bert2lstm.py:115]: I must have commented those lines out while testing. I will uncomment those

Comment (Het-Shah) [reply KD_Lib/KD/text/Bert2Lstm/bert2lstm.py:115]: cool

**Comment by Het-Shah** [comment]
> 1. I am not very familiar with tensorboard and hence did not include it here. I would request someone else to possibly take this up or maybe I can do it later.

I think this is simple enough to learn and it is going to be an essential thing in projects that you may work other than this. 

> 2. That seems like a nice approach. Are you suggesting that we have a class for the model which handles its training/distillation, etc.?

Yes so basically all the functions related to, lets say bert go into one class. 

I think after these changes we should be good with this implementation.

**Comment by avishreekh** [comment]
> Yea seems a bit better but the size of the file is really huge. We will need to figure a way out to reduce that.

We would then need to write the training, distillation functions separately

**Comment by Het-Shah** [comment]
> We would then need to write the training, distillation functions separately

Yea I think we will have to figure out some way to do this efficiently.

**Comment by avishreekh** [comment]
> > We would then need to write the training, distillation functions separately
> 
> Yea I think we will have to figure out some way to do this efficiently.

I think we have the following options:
1. Create handler classes for LSTM and BERT
2. Create handler functions for LSTM and BERT
3. Derive a class from the base class for distilling to LSTM and then derive Bert2LSTM from that
Let me know your thoughts on this

**Comment by avishreekh** [comment]
Maybe we can sort this later when adding new implementations using these two models. That would give us more perspective on how to structure them

**Comment by Het-Shah** [comment]
> I think we have the following options:
> 
> 1. Create handler classes for LSTM and BERT
> 2. Create handler functions for LSTM and BERT
> 3. Derive a class from the base class for distilling to LSTM and then derive Bert2LSTM from that
>    Let me know your thoughts on this

I really like this approach will be good for a start.

**Comment by Het-Shah** [comment]
> Maybe we can sort this later when adding new implementations using these two models. That would give us more perspective on how to structure them

Yea lets just get this implementation to work as of now, we will try to clean it up in another pr

File: logs.rst [hunk 3]
```diff
@@ -51,10 +51,9 @@ Results for Knowledge Distillation Algorithms -
...
 +------------------------+----------------------------------+------------------------------+--------------------------------+----------------------+---------------------+
-|  BERT2LSTM             |               IMDB               |  BERT(base-uncased) - Teacher|                                |                      |     0.8366          |
+|  BERT2LSTM             |               IMDB               |  BERT(base-uncased) - Teacher|            109483778           |                      |     0.8366          |
 |                        |                                  +------------------------------+--------------------------------+         5            +---------------------+
-|                        |                                  |  LSTM - Student          	   |                                |                      |     0.          |
-|                        |                                  |  			            	   |                                |                      |     0.          |
+|                        |                                  |  LSTM - Student          	   |             1051274            |                      |     0.8372          |
```
[the remark this answers is not in the document]
Comment (avishreekh) [reply logs.rst]: Oops :smile:

**Comment by Het-Shah** [comment]
Did you pull? I made changes to `README.rst`. I'll merge it after you pull the master branch.

**Review by Het-Shah — APPROVED** [review]
Looks good. Merging
